### DISCUSSION TITLE
TShock 4.5.0 for Terraria 1.4.2.1

### DISCUSSION THREAD

**Discussion by hakusaro** [Pryaxis/TShock#2268] [Show and tell]
## Changes
* Updated OTAPI and TSAPI to Terraria 1.4.2.1. (@Stealownz, @DeathCradle)
* Updated TShock with preliminary protocol support for Terraria 1.4.2.1. (@Stealownz)

**Thank you so much to @Stealownz and @DeathCradle for this release!**

## Notes
An executive decision has been made on versioning going forward. Instead of publishing pre-releases on 4.4.0 forever, this is now 4.5.0. Patch releases on the 4.5 series will be released as time goes on. You should consider the 4.5.x series unstable. All _odd numbered_ versions (4.5, 4.7, 4.9) will be unstable. When a stable version is released, it will be even (e.g., 4.6.0, 4.8.0, 6.0.0).

The reason for this change is simple. The TShock automatic update system does not display updates for pre-releases. Unfortunately, it is quite useless to have an automatic updater when you don't use it. Polishing 4.4.0 has been time consuming, and therefore, we haven't been able to push a stable version for a while. Because of this, the lack of an automatic update function means that people are left in the dark about important updates. This is a problem.

We are planning on publishing updates more frequently to resolve stability issues and improve the system. Unfortunately, it's just taking a bit longer than usual.

To upgrade, simply overwrite all applicable files in your existing server folder. TShock releases are designed so that you can upgrade by copying all files over and clicking "yes" to overwrite any duplicates. All existing world and character and database data will be saved. **Please make a backup copy of your world file and your TShock configuration before installing**.

<hr /><em>This discussion was created from the release <a href='https://github.com/Pryaxis/TShock/releases/tag/v4.5.0'>TShock 4.5.0 for Terraria 1.4.2.1</a>.</em>

**Comment by ATFGK** [comment]
Cannot run when using MySQL database...

**Reply by ATFGK** [reply]
2021-04-11 14:25:52 - TShock: ERROR: Fatal Startup Exception
2021-04-11 14:25:52 - TShock: ERROR: MySql.Data.MySqlClient.MySqlException (0x80004005): Unknown column 'name' in 'field list'
   at MySql.Data.MySqlClient.MySqlStream.ReadPacket()
   at MySql.Data.MySqlClient.NativeDriver.GetResult(Int32& affectedRow, Int64& insertedId)
   at MySql.Data.MySqlClient.Driver.GetResult(Int32 statementId, Int32& affectedRows, Int64& insertedId)
   at MySql.Data.MySqlClient.Driver.NextResult(Int32 statementId, Boolean force)
   at MySql.Data.MySqlClient.MySqlDataReader.NextResult()
   at MySql.Data.MySqlClient.MySqlCommand.ExecuteReader(CommandBehavior behavior)
   at MySql.Data.MySqlClient.MySqlCommand.ExecuteScalar()
   at TShockAPI.DB.DbExt.QueryScalar[T](IDbConnection olddb, String query, Object[] args)
   at TShockAPI.DB.BanManager.TryConvertBans()
   at TShockAPI.DB.BanManager..ctor(IDbConnection db)
   at TShockAPI.TShock.Initialize()

**Reply by hakusaro** [reply]
That's not good! We'll have to look into that.

**Reply by hakusaro** [reply]
@ATFGK can you drop by discord or test with the unstable build from https://github.com/Pryaxis/TShock/pull/2269/checks?check_run_id=2316118710? **Make a copy of your ban database first please!**

**Reply by ATFGK** [reply]
> @ATFGK can you drop by discord or test with the unstable build from https://github.com/Pryaxis/TShock/pull/2269/checks?check_run_id=2316118710? **Make a copy of your ban database first please!**

I use a brand new database to test and still can't run

**Reply by ATFGK** [reply]
> > @ATFGK can you drop by discord or test with the unstable build from https://github.com/Pryaxis/TShock/pull/2269/checks?check_run_id=2316118710? **Make a copy of your ban database first please!**
> 
> I use a brand new database to test and still can't run

only set up  MySQL in config.json

**Reply by hakusaro** [reply]
With the latest build from that pull request?

**Reply by hakusaro** [reply]
https://github.com/Pryaxis/TShock/suites/2467558078/artifacts/53042154

**Reply by ATFGK** [reply]
> https://github.com/Pryaxis/TShock/suites/2467558078/artifacts/53042154
It worked, but "bans" in the database disappeared

**Comment by Kojirremer** [comment]
```
2021-04-11 01:40:32 - TShock: ERROR: Fatal Startup Exception
2021-04-11 01:40:32 - TShock: ERROR: Mono.Data.Sqlite.SqliteException (0x80004005): The database file is locked
database is locked
   at Mono.Data.Sqlite.SQLite3.Step(SqliteStatement stmt)
   at Mono.Data.Sqlite.SqliteDataReader.NextResult()
   at Mono.Data.Sqlite.SqliteDataReader..ctor(SqliteCommand cmd, CommandBehavior behave)
   at Mono.Data.Sqlite.SqliteCommand.ExecuteReader(CommandBehavior behavior)
   at Mono.Data.Sqlite.SqliteCommand.ExecuteScalar()
   at TShockAPI.DB.DbExt.QueryScalar[T](IDbConnection olddb, String query, Object[] args)
   at TShockAPI.DB.BanManager.InsertBan(String identifier, String reason, String banningUser, DateTime fromDate, DateTime toDate)
   at TShockAPI.DB.BanManager.TryConvertBans()
   at TShockAPI.DB.BanManager..ctor(IDbConnection db)
   at TShockAPI.TShock.Initialize()
```
   I'm getting this on startup. Console doesn't even give an error message. It just stops shortly after opening, then console closes after a few seconds. No idea what is causing it as I tried starting up without any plugins other than TShockAPI.dll and it still crashes. Maybe old files are clashing with new ones. I don't really know.
Will try on a (mostly) clean installation tomorrow.

**Reply by hakusaro** [reply]
Is some other process using the database?

**Reply by Kojirremer** [reply]
> 
> 
> Is some other process using the database?

Not that I know of. I tried both on the server and desktop and get the same result. Also checked that tshock.sqlite wasn't set as Read-Only. I though maybe one of those plugins that put their own tables within tshock.sqlite may have been causing it, but running it without a single plugin gives the same result.
I'll check on a clean install tomorrow. I have a feeling leftover .dlls from older versions are starting to have conflicts with new ones, as some files were present in both root and plugins folder.

**Reply by Kojirremer** [reply]
I tried deleting all entries on the Bans table on tshock.sqlite and the server finally managed to start, so the problem is probably around that table.

**Comment by Marcus101RR** [comment]
![image](https://user-images.githubusercontent.com/8344293/114315748-4f576480-9ace-11eb-9030-e3fbc2a44955.png)
Since when did this get changed and where was the documentation on it?

**Reply by Terrabade** [reply]
Appears the only "documentation" was the error message, I checked and it looks like it was done early last December.

**Reply by Marcus101RR** [reply]
To anyone looking for the answer, add "Settings." to fix this issue.

**Reply by QuiCM** [reply]
Agreed communication on this could have been better. I'll endeavour to document/communicate breaking changes like this better in future

**Comment by ATFGK** [comment]
Wrong word in "public bool Ban(string reason, bool force = false, string adminUserName = null)">TShock.Bans.InsertBan($"**{Identifier.IP}**{UUID}",

**Reply by hakusaro** [reply]
Thanks for reporting this!

**Comment by ATFGK** [comment]
When "RestApiEnabled:true", the server crashes and prompts "Process is terminated due to StackOverflowException."

**Reply by QuiCM** [reply]
Hi,
I can't reproduce the same issue. Do you have any specific REST settings defined?

**Reply by ATFGK** [reply]
Only use "v2/players/list" when there are players online

**Reply by ATFGK** [reply]
No plug-ins were installed at that time

**Reply by ATFGK** [reply]
It's probably the problem in these places: 
Rest.cs>
Enum.TryParse(rObj.Status, out HttpStatusCode status)
RestManager.cs>
var playerList = new ArrayList();
...
return new RestObject() { { "players", playerList } };

**Reply by QuiCM** [reply]
Thanks. I've fixed this in https://github.com/Pryaxis/TShock/commit/32a0e2f1b6474632daaba3ccef757760c6e54576

**Reply by ATFGK** [reply]
> Thanks. I've fixed this in [32a0e2f](https://github.com/Pryaxis/TShock/commit/32a0e2f1b6474632daaba3ccef757760c6e54576)

Thanks a lot!!!!
